File: app/src/main/java/com/prox/voicechanger/recorder/Recorder.java
package com.prox.voicechanger.recorder;

import static com.prox.voicechanger.VoiceChangerApp.TAG;

import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.util.Log;

import com.prox.voicechanger.interfaces.RecorderListener;
import com.prox.voicechanger.utils.FileUtils;

import java.io.IOException;

public class Recorder implements RecorderListener {
    private static final int SAMPLING_RATE_IN_HZ = 8000;
    private static final int CHANNEL_CONFIG = AudioFormat.CHANNEL_IN_MONO;
    private static final int CHANNEL_COUNT = 1;
    private static final int AUDIO_FORMAT = AudioFormat.ENCODING_PCM_16BIT;
    private static final int BIT_PER_SAMPLE = 16;
    private static final long BYTE_RATE = BIT_PER_SAMPLE * SAMPLING_RATE_IN_HZ * CHANNEL_COUNT / 8;
    private static final int BUFFER_SIZE = AudioRecord.getMinBufferSize(SAMPLING_RATE_IN_HZ, CHANNEL_CONFIG, AUDIO_FORMAT);

    private MediaRecorder recorder;
    private String path;
    private String name;
    private long startTime;
    private boolean isRecording;

    public Recorder() {
        path = FileUtils.getRecordingFilePath();
        name = FileUtils.getName(path);
        Log.d(TAG, "Recorder: create "+path);

        recorder = new MediaRecorder();
        recorder.setAudioSource(MediaRecorder.AudioSource.MIC);
        recorder.setOutputFormat(MediaRecorder.OutputFormat.THREE_GPP);
        recorder.setOutputFile(path);
        recorder.setAudioEncoder(MediaRecorder.AudioEncoder.AMR_NB);
        startTime = 0;
        isRecording = false;
    }

    @Override
    public void start() {
        if (recorder==null){
            Log.d(TAG, "Recorder: start null");
            return;
        }
        if (!isRecording){
            try {
                recorder.prepare();
                recorder.start();
                startTime = System.currentTimeMillis();
                isRecording = true;
                Log.d(TAG, "Recorder: start");
            } catch (IOException e) {
                Log.d(TAG, "Recorder: error start - "+e.getMessage());
            }
        }
    }

    @Override
    public void stop() {
        if (recorder==null){
            Log.d(TAG, "Recorder: stop null");
            return;
        }
        if (isRecording){
            recorder.stop();
            recorder.release();
            startTime = 0;
            isRecording = false;
            Log.d(TAG, "Recorder: stop");
        }
    }

    @Override
    public void release() {
        if (recorder==null){
            Log.d(TAG, "Recorder: release null");
            return;
        }
        if (isRecording){
            recorder.stop();
        }
        recorder.release();
        recorder = null;
        path = null;
        name = null;
        startTime = 0;
        isRecording = false;
        Log.d(TAG, "Recorder: release");
    }

    public long getCurrentTime() {
        return System.currentTimeMillis() - startTime;
    }

    public int getTickDuration() {
        return (int)((double)BUFFER_SIZE * 500 / (BYTE_RATE));
    }

    public int getMaxAmplitude() {
        return recorder.getMaxAmplitude();
    }

    public String getPath() {
        return path;
    }
    public String getName() {
        return name;
    }
}

File: app/src/main/java/com/prox/voicechanger/media/Recorder.kt
package com.prox.voicechanger.media

import android.content.Context
import android.media.*
import com.prox.voicechanger.utils.FileUtils.Companion.getTempRecordingFilePath
import com.prox.voicechanger.VoiceChangerApp
import com.prox.voicechanger.interfaces.RecorderListener
import android.util.Log
import java.io.IOException

class Recorder(context: Context?) : RecorderListener {
    private var recorder: MediaRecorder?
    var path: String?
        private set
    private var startTime: Long
    private var isRecording: Boolean
    override fun start() {
        if (recorder == null) {
            Log.d(VoiceChangerApp.TAG, "Recorder: start null")
            return
        }
        if (!isRecording) {
            try {
                recorder!!.prepare()
                recorder!!.start()
                startTime = System.currentTimeMillis()
                isRecording = true
                Log.d(VoiceChangerApp.TAG, "Recorder: start")
            } catch (e: IOException) {
                Log.d(VoiceChangerApp.TAG, "Recorder: error start - " + e.message)
            }
        }
    }

    override fun stop() {
        if (recorder == null) {
            Log.d(VoiceChangerApp.TAG, "Recorder: stop null")
            return
        }
        if (isRecording) {
            recorder!!.stop()
            recorder!!.release()
            startTime = 0
            isRecording = false
            Log.d(VoiceChangerApp.TAG, "Recorder: stop")
        }
    }

    override fun release() {
        if (recorder == null) {
            Log.d(VoiceChangerApp.TAG, "Recorder: release null")
            return
        }
        if (isRecording) {
            recorder!!.stop()
        }
        recorder!!.release()
        recorder = null
        path = null
        startTime = 0
        isRecording = false
        Log.d(VoiceChangerApp.TAG, "Recorder: release")
    }

    val currentTime: Long
        get() = System.currentTimeMillis() - startTime
    val tickDuration: Int
        get() = (BUFFER_SIZE.toDouble() * 500 / BYTE_RATE).toInt()
    val maxAmplitude: Int
        get() = if (recorder != null) {
            recorder!!.maxAmplitude
        } else 0

    companion object {
        private const val SAMPLING_RATE_IN_HZ = 8000
        private const val CHANNEL_CONFIG = AudioFormat.CHANNEL_IN_MONO
        private const val CHANNEL_COUNT = 1
        private const val AUDIO_FORMAT = AudioFormat.ENCODING_PCM_16BIT
        private const val BIT_PER_SAMPLE = 16
        private const val BYTE_RATE =
            (BIT_PER_SAMPLE * SAMPLING_RATE_IN_HZ * CHANNEL_COUNT / 8).toLong()
        private val BUFFER_SIZE =
            AudioRecord.getMinBufferSize(SAMPLING_RATE_IN_HZ, CHANNEL_CONFIG, AUDIO_FORMAT)
    }

    init {
        path = getTempRecordingFilePath(context!!)
        Log.d(VoiceChangerApp.TAG, "Recorder: create $path")
        recorder = MediaRecorder()
        recorder!!.setAudioSource(MediaRecorder.AudioSource.MIC)
        recorder!!.setOutputFormat(MediaRecorder.OutputFormat.THREE_GPP)
        recorder!!.setOutputFile(path)
        recorder!!.setAudioEncoder(MediaRecorder.AudioEncoder.AMR_NB)
        startTime = 0
        isRecording = false
    }
}